lib/AST/ReachBasedRefine.cpp

```diff
@@ -40,83 +26,116 @@ z3::expr ReachBasedRefine::GetZ3Cond(clang::IfStmt *ifstmt) {
-    // conditions of all the IfStmts form a tautology.
-    if (IsTautology()) {
-      break;
+bool ReachBasedRefine::VisitCompoundStmt(clang::CompoundStmt *compound) {
+  std::vector<clang::Stmt *> body{compound->body_begin(), compound->body_end()};
+  std::vector<clang::IfStmt *> ifs;
+  z3::expr_vector conds{*z3_ctx};
+
+  bool done_something{false};
+  for (size_t i{0}; i < body.size(); ++i) {
+    auto if_stmt{clang::dyn_cast<clang::IfStmt>(body[i])};
+    if (!if_stmt) {
+      ifs.clear();
+      conds.resize(0);
+      continue;
     }
-    // Clear else-if IfStmts if we find a path among them.
-    auto cond = GetZ3Cond(stmt);
-    if (stmt->getElse() || !IsUnrechable(cond)) {
-      conds = z3::expr_vector(*z3_ctx);
-      elifs.clear();
+
+    ifs.push_back(if_stmt);
+    auto cond{GetZ3Cond(if_stmt)};
+
+    if (if_stmt->getElse()) {
+      // We cannot link `if` statements that contain `else` branches
+      ifs.clear();
+      conds.resize(0);
+      continue;
     }
-    // Add the current if-statement to the else-if candidates.
-    conds.push_back(cond);
-    elifs.push_back(stmt);
-  }
 
-  // Check if we have enough statements to work with
-  if (elifs.size() < 2) {
-    return;
-  }
+    // Is the current `if` statement unreachable from all the others?
+    bool is_unreachable{Prove(*z3_ctx, !(cond && z3::mk_or(conds)))};
 
-  // Create the else-if cascade
-  clang::IfStmt *sub = nullptr;
-  for (auto stmt : llvm::make_range(elifs.rbegin(), elifs.rend())) {
-    auto cond = stmt->getCond();
-    auto then = stmt->getThen();
-    if (stmt == elifs.back()) {
-      sub = ast.CreateIf(cond, then);
-      substitutions[stmt] = sub;
-    } else if (stmt == elifs.front()) {
-      std::vector<clang::Stmt *> thens({then});
-      sub->setElse(ast.CreateCompoundStmt(thens));
-      substitutions[stmt] = nullptr;
-    } else {
-      auto elif = ast.CreateIf(cond, then);
-      sub->setElse(elif);
-      sub = elif;
-      substitutions[stmt] = nullptr;
+    if (!is_unreachable) {
+      ifs.clear();
+      conds.resize(0);
+      continue;
     }
-  }
-}
 
-bool ReachBasedRefine::VisitCompoundStmt(clang::CompoundStmt *compound) {
-  // DLOG(INFO) << "VisitCompoundStmt";
-  // Create else-if cascade substitutions for IfStmts in `compound`
-  CreateIfElseStmts(GetIfStmts(compound));
-  // Apply created else-if substitutions and
-  // create a replacement for `compound`
-  if (ReplaceChildren(compound, substitutions)) {
-    std::vector<clang::Stmt *> new_body;
-    for (auto stmt : compound->body()) {
-      if (stmt) {
-        new_body.push_back(stmt);
+    conds.push_back(GetZ3Cond(if_stmt));
+
+    // Do the collected statements cover all possibilities?
+    auto is_complete{Prove(*z3_ctx, z3::mk_or(conds))};
+
+    if (ifs.size() <= 2 || !is_complete) {
+      // We need to collect more statements
+      continue;
+    }
+
+    /*
+    `body` will look like this at this point:
```

Comment:
This is my subjective preference so take it with a pinch of salt, but I usually prefer comments to be short / dense so that it doesn't break the flow of the code and a relevant scope or function is on one editor screen. That being said, the comment is excellent so it'd be a shame to change it.

Reply:
Yeah I figured the best way to show what the code was doing was a little "picture", since indices are confusing otherwise